### PR TITLE
Renewable output [3/3]

### LINKED ISSUE #220: Feature wind off-shore (or miscellaneus renewable)
This is a new feature about a new kind of renewable (could be called misc. renewable).
More accurately, we want to offer the possibility to add renewable clusters (of wind or solar type) to any area and provide many time series (or TS) to each renewable cluster. For each renewable cluster and a given year, the related ready-made TS number is chosen by the user or drawn randomly.
 
## Antares main Window
- A new tab is to be added in the main notebook, named Renewable (or misc. renewable). 
- This tab is built on the model of the Thermal tab : it contains an area selector and a possibility to **add**/**delete** (possibly **clone**) renewable clusters to any area.
- Once an area is selected (as well as a cluster), 3 frames must appear : 
   - the list of renewable clusters (**up**),
   - a minimal frame containing some characteristics of the selected cluster : name, type [wind or solar], area it belongs to (**left**)
   - a table containing a time series per column (**right**)

### "Old" and "New" renewables

The existent **Wind** and **Solar** tabs are kept as they are now. To avoid confusion, a toggle button is to be added to select between "old" and "new" renewable.

Old renewables allow TS generation, but only "Solar" and "Wind".

New renewables do not allow TS generation, but multiple types of solar & wind.

## Scenario builder
The scenario builder (choice of TS number for a year and a cluster) is similar to the thermal scenario builder.

## Input
A new directory is created in the **input** directory, called **renewables**, with the same structure as the **thermal** directory, except for the **prepro** subdirectory, nonexistant here.
We have to read the content of this directory when we load the study, and write it when some cluster data changed or when we make a **save as** operation. 

### Building outputs
Regarding the renewable clusters, there are 2 types of results : 
- The classical **Wind** and **Solar** results. These results can be found only in the result files whose names are "values" prefixed. To the classical results (the TS chosen randomly or by user), we need to add the TS chosen for the renewable clusters of the same type : for example, for a given area, we add to the **Wind** classical results, the chosen TS associated to all renewable clusters of type **Wind** in the area.
So we only have to make an addition to variable **Wind** and **Solar** before they are printed. Should be short and easy.
- A new kind of results : the details of each renewable clusters, as it is done for the thermal clusters. Of course, these results are based on the TS chosen before the simulation, whether they were chosen by user (via scenario builder) or randomly. 

### Reading outputs
- Tab **General values**
Reading these file is already done. Nothong to do.
- Tab **Renewable values**
This should be a new tab, close to the **Thermal plants** tab and similar to it : it contains the details of every renewable cluster.

## Solver
- Data structure : we provide a data structure to contain the production of each renewable cluster.
**Caution** : we may have to refactor the already existing cluster class, so that it contains a common part (parent class),  a  thermal cluster part (derived from the parent class), and a renewable cluster class (also derived from the parent class).
- In case the scenario builder is not used for any renewable cluster, we have to choose randomly which TS is associated to that cluster.
-  As it is already done for **Wind** and **Solar**,  we must substract from the load, the production of the renewable clusters

### PR DESCRIPTION
**Pull request by flomnes**
Write renewable input data into the `output` directory 
* year by year
* synthesis
* record years

For each time granularity
* hourly
* daily
* weekly
* monthly
* annual

### CODE REVIEW

File: src/solver/variable/area.inc.hxx [hunk 1]
```diff
@@ -241,6 +250,16 @@ void Areas<NEXTTYPE>::yearEndBuild(State& state, uint year, uint numSpace)
             // Variables
             variablesForArea.yearEndBuildForEachThermalCluster(state, year, numSpace);
         } // for each thermal cluster
+
+        // For each renewable cluster
+        for (uint j = 0; j != area.renewable.clusterCount; ++j)
+        {
+            state.renewableCluster = area.renewable.clusters[j];
+            state.yearEndReset();
```
Comment (guilpier-code) [inline src/solver/variable/area.inc.hxx]: **yearEndReset** has already been called for thermal clusters. Calling it here again could set to zero some results computed for thermal clusters.

File: src/solver/variable/area.inc.hxx [hunk 2]
```diff
@@ -229,7 +238,7 @@ void Areas<NEXTTYPE>::yearEndBuild(State& state, uint year, uint numSpace)
         // For each thermal cluster
         for (uint j = 0; j != area.thermal.clusterCount; ++j)
         {
-            state.cluster = area.thermal.clusters[j];
+            state.thermalCluster = area.thermal.clusters[j];
             state.yearEndReset();
 
             // Variables
```
Comment (guilpier-code) [inline src/solver/variable/area.inc.hxx:244]: **yearEndBuildFromThermalClusterIndex** function name is very unclear on what the function really does.
This name should be changed, even if it has nothing to see with the current PR.

File: src/solver/variable/commons/links/links.hxx.inc.hxx [hunk 3]
```diff
@@ -104,6 +104,21 @@ inline void Links::yearEndBuildPrepareDataForEachThermalCluster(State& state,
     }
 }
 
+inline void Links::yearEndBuildPrepareDataForEachRenewableCluster(State& state,
```
Comment (guilpier-code) [inline src/solver/variable/commons/links/links.hxx.inc.hxx:107]: This function is useless for now, but might be used someday (to be checked).

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/solver/variable/commons/links/links.hxx.inc.hxx (hunk 3 above)
Comment (flomnes) [reply src/solver/variable/commons/links/links.hxx.inc.hxx:107]: This function is used to recover hourly data for renewable clusters, see **src/solver/variable/economy/renewableGeneration.h:290**.
```cpp
void yearEndBuildPrepareDataForEachRenewableCluster(State& state,
                                                    uint year,
                                                    unsigned int numSpace)
{
    for (unsigned int i = 0; i <= state.study.runtime->rangeLimits.hour[Data::rangeEnd]; ++i)
    {
        state.renewableClusterProductionForYear[i]
          += pValuesForTheCurrentYear[numSpace][state.renewableCluster->areaWideIndex].hour[i];
    }
 // Next variable
    NextType::yearEndBuildPrepareDataForEachRenewableCluster(state, year, numSpace);
}
```

File: src/solver/variable/area.inc.hxx (hunk 2 above)
Comment (flomnes) [reply src/solver/variable/area.inc.hxx:244]: How about **yearEndRecoverThermalData** ?

File: src/solver/variable/area.inc.hxx (hunk 1 above)
Comment (flomnes) [reply src/solver/variable/area.inc.hxx]: I duplicated this function into two : one for thermal data, one for renewable data. See commit 9573759.

**Comment by flomnes** [comment]
Local merge in PR #267

**Pull request closed without merging**